File: web_app.py
"""Small local web UI for the seven-period Gunn room highlighter."""

from __future__ import annotations

import json
import os
import re
import time
import uuid
from collections import defaultdict
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import urlparse

from map_highlighter import highlight_rooms


ROOT = Path(__file__).resolve().parent
WEB = ROOT / "web"
MAP = ROOT / "src" / "map" / "gunn_site_map.png"
OUTPUT = ROOT / "output" / "period_map.png"
ROOMS = json.loads((ROOT / "room_regions.json").read_text(encoding="utf-8"))["rooms"]
BUILDINGS = sorted({room["building"] for room in ROOMS}, key=lambda name: (name == "BG", name))
BY_ID = {room["id"].casefold(): room for room in ROOMS}
BY_BUILDING = defaultdict(list)
for _room in ROOMS:
    BY_BUILDING[_room["building"]].append(_room)


def _normalise(value: str) -> str:
    return re.sub(r"[\s-]+", "", value).casefold()


def _resolve_room(building: str, value: str) -> dict:
    if building not in BY_BUILDING:
        raise ValueError(f"Unknown building: {building}")
    value = value.strip()
    tagged = re.fullmatch(r".+\((R\d{3})\)", value, flags=re.IGNORECASE)
    key = (tagged.group(1) if tagged else value).casefold()
    if key in BY_ID:
        room = BY_ID[key]
        if room["building"] != building:
            raise ValueError(f"{room['label']} is not in {building} Building")
        return room

    matches = [
        room
        for room in BY_BUILDING[building]
        if _normalise(value) in {_normalise(room["label"]), *(_normalise(alias) for alias in room.get("aliases", []))}
    ]
    if not matches:
        raise ValueError(f"Room {value!r} was not found in {building} Building")
    if len(matches) > 1:
        choices = "、".join(f"{room['label']} ({room['id']})" for room in matches)
        raise ValueError(f"{value} appears twice on the map. Choose {choices}")
    return matches[0]


def _render_periods(periods: object) -> dict:
    if not isinstance(periods, list) or len(periods) != 7:
        raise ValueError("Please submit all seven period slots")
    colors = {}
    selected = []
    warnings = []
    seen = {}
    for index, period in enumerate(periods, 1):
        if not isinstance(period, dict):
            raise ValueError(f"Period {index} has invalid data")
        building = str(period.get("building", "")).strip().upper()
        room_name = str(period.get("room", "")).strip()
        color = str(period.get("color", "")).strip()
        if not building and not room_name:
            continue
        if not building or not room_name:
            raise ValueError(f"Period {index}: choose a building and enter a room")
        if not re.fullmatch(r"#[0-9a-fA-F]{6}", color):
            raise ValueError(f"Period {index}: invalid color")
        try:
            room = _resolve_room(building, room_name)
        except ValueError as exc:
            raise ValueError(f"Period {index}: {exc}") from exc
        if room["id"] in seen:
            warnings.append(f"Periods {seen[room['id']]} and {index} both use {room['label']}; the map uses Period {index}'s color.")
        seen[room["id"]] = index
        colors[room["id"]] = color
        selected.append({"period": index, "id": room["id"], "label": room["label"], "building": building, "floor": room.get("floor", 1), "color": color})

    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    temp = OUTPUT.parent / f".period_map_{uuid.uuid4().hex}.png"
    try:
        highlight_rooms(colors, temp, opacity=0.55)
        os.replace(temp, OUTPUT)
    finally:
        temp.unlink(missing_ok=True)
    return {"image_url": f"/output/period_map.png?v={time.time_ns()}", "selected": selected, "warnings": warnings}


class Handler(BaseHTTPRequestHandler):
    def _send(self, status: int, body: bytes, content_type: str) -> None:
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(body)

    def _file(self, path: Path, content_type: str) -> None:
        if not path.is_file():
            self.send_error(404)
            return
        self._send(200, path.read_bytes(), content_type)

    def do_GET(self) -> None:
        path = urlparse(self.path).path
        if path == "/":
            return self._file(WEB / "index.html", "text/html; charset=utf-8")
        if path == "/app.js":
            return self._file(WEB / "app.js", "text/javascript; charset=utf-8")
        if path == "/style.css":
            return self._file(WEB / "style.css", "text/css; charset=utf-8")
        if path == "/map.png":
            return self._file(MAP, "image/png")
        if path == "/output/period_map.png":
            return self._file(OUTPUT, "image/png")
        if path == "/api/rooms":
            data = {
                "buildings": BUILDINGS,
                "rooms": [
                    {"id": room["id"], "label": room["label"], "building": room["building"], "floor": room.get("floor", 1), "aliases": room.get("aliases", [])}
                    for room in ROOMS
                ],
            }
            return self._send(200, json.dumps(data, ensure_ascii=False).encode("utf-8"), "application/json; charset=utf-8")
        if path == "/favicon.ico":
            return self._send(204, b"", "image/x-icon")
        self.send_error(404)

    def do_POST(self) -> None:
        if urlparse(self.path).path != "/api/render":
            self.send_error(404)
            return
        try:
            length = int(self.headers.get("Content-Length", "0"))
            if not 0 < length <= 16000:
                raise ValueError("Request is empty or too large")
            payload = json.loads(self.rfile.read(length))
            result = _render_periods(payload.get("periods"))
        except (ValueError, TypeError, AttributeError, json.JSONDecodeError) as exc:
            body = json.dumps({"error": str(exc)}, ensure_ascii=False).encode("utf-8")
            return self._send(400, body, "application/json; charset=utf-8")
        body = json.dumps(result, ensure_ascii=False).encode("utf-8")
        self._send(200, body, "application/json; charset=utf-8")


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--port", type=int, default=8000)
    args = parser.parse_args()
    server = ThreadingHTTPServer(("127.0.0.1", args.port), Handler)
    print(f"Open http://127.0.0.1:{args.port}/", flush=True)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()

File: web_app.ts
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { randomUUID } from 'node:crypto';
import { mkdir, readFile, rename, writeFile, rm } from 'node:fs/promises';
import { resolve } from 'node:path';
import { pathToFileURL } from 'node:url';
import sharp from 'sharp';
import { ROOT, rooms, buildings, roomData, resolveRoom } from './project.js';
import { evacuationForRoom } from './evacuation.js';
import { renderRooms, svg, xml } from './map_highlighter.js';
export { resolveRoom } from './project.js';
class InputError extends Error {}
interface LegendItem { period: number; label: string; floor: number; color: string }
async function addScheduleLegend(image: Buffer, selected: LegendItem[]): Promise<Buffer> {
  if (!selected.length) return image;
  const { width, height } = await sharp(image).metadata();
  if (!width || !height) throw new Error('Map dimensions are missing');
  // Keep the original canvas size so room polygons and clickable markers stay aligned.
  const padding = 24, headerHeight = 42, rowHeight = 34, legendWidth = 420;
  const legendHeight = padding * 2 + headerHeight + rowHeight * selected.length;
  const left = Math.max(padding, Math.min(420, width - legendWidth - padding));
  const top = height - legendHeight - padding;
  const rows = selected.map((item, index) => {
    const y = top + padding + headerHeight + index * rowHeight;
    const label = `Period ${item.period} · ${item.label}${item.floor === 2 ? ' (2F)' : ''}`;
    return `<rect x="${left + padding}" y="${y + 4}" width="22" height="22" fill="${xml(item.color)}"/><text x="${left + padding + 34}" y="${y + 22}" font-size="22" fill="#344054">${xml(label)}</text>`;
  }).join('');
  const body = `<g font-family="sans-serif"><rect x="${left}" y="${top}" width="${legendWidth}" height="${legendHeight}" fill="white" stroke="#d0d5dd" stroke-width="2"/><text x="${left + padding}" y="${top + padding + 26}" font-size="26" fill="#182334">Schedule Legend</text>${rows}</g>`;
  return sharp(image).composite([{ input: svg(width, height, body) }]).removeAlpha().png().toBuffer();
}
export async function renderPeriods(periods: unknown, outputDir = resolve(ROOT,'output')) {
  if (!Array.isArray(periods) || periods.length !== 7) throw new InputError('Please submit all seven period slots');
  const colors: Record<string,string[]> = {};
  const selected = [], warnings: string[] = [];
  for (const [i,p] of periods.entries()) {
    const index=i+1;
    if (!p || typeof p !== 'object' || Array.isArray(p)) throw new InputError(`Period ${index} has invalid data`);
    const building=String(p.building??'').trim().toUpperCase(), roomName=String(p.room??'').trim(), color=String(p.color??'').trim();
    if (!building && !roomName) continue;
    if (!building || !roomName) throw new InputError(`Period ${index}: choose a building and enter a room`);
    if (!/^#[0-9a-f]{6}$/i.test(color)) throw new InputError(`Period ${index}: invalid color`);
    let room;
    try {room=resolveRoom(building,roomName);} catch(error) {throw new InputError(`Period ${index}: ${(error as Error).message}`);}
    (colors[room.id] ??= []).push(color);
    selected.push({period:index,id:room.id,label:room.label,building,floor:room.floor??1,color,polygon:room.polygon,
      evacuation:evacuationForRoom(room),marker:[(room.label_box[0]+room.label_box[2])/2,(room.label_box[1]+room.label_box[3])/2]});
  }
  for (const id of Object.keys(colors)) {
    const shared = selected.filter(item => item.id === id);
    if (shared.length > 1) warnings.push(`Periods ${shared.map(item => item.period).join(', ')} share ${shared[0].label}; its map highlight is split into each period's color.`);
  }
  const bytes=await addScheduleLegend(await renderRooms(colors,{opacity:0.55}),selected);
  await mkdir(outputDir,{recursive:true});
  const id=randomUUID().replaceAll('-',''), filename=`period_map_${id}.png`, latest=resolve(outputDir,`.latest_map_${id}.png`);
  await writeFile(resolve(outputDir,filename),bytes,{flag:'wx'});
  try {await writeFile(latest,bytes); await rename(latest,resolve(outputDir,'period_map.png'));}
  finally {await rm(latest,{force:true});}
  return {image_url:`/output/${filename}`,selected,warnings,map_size:roomData.image_size};
}
function send(res:ServerResponse,status:number,body:string|Buffer,type='application/json; charset=utf-8') {
  res.writeHead(status,{'Content-Type':type,'Content-Length':Buffer.byteLength(body),'Cache-Control':'no-store'});res.end(body);
}
async function payload(req:IncomingMessage):Promise<unknown> {
  const chunks:Buffer[]=[]; let size=0;
  for await (const chunk of req) {
    size+=chunk.length;
    if (size>16000) throw new InputError('Request is empty or too large');
    chunks.push(chunk);
  }
  if (!size) throw new InputError('Request is empty or too large');
  try {return JSON.parse(Buffer.concat(chunks).toString('utf8'));}
  catch {throw new InputError('Invalid JSON request');}
}
export function createApp(outputDir=resolve(ROOT,'output')) {
  return createServer(async(req,res)=>{
    try {
      const pathname=new URL(req.url??'/', 'http://localhost').pathname;
      if(req.method==='POST' && pathname==='/api/render') {
        const body=await payload(req);
        if(!body || typeof body!=='object' || Array.isArray(body)) throw new InputError('Invalid request object');
        return send(res,200,JSON.stringify(await renderPeriods((body as {periods?:unknown}).periods,outputDir)));
      }
      if(req.method!=='GET') return send(res,req.method==='POST'?404:405,JSON.stringify({error:'Not found'}));
      if(pathname==='/api/rooms') return send(res,200,JSON.stringify({buildings,rooms:rooms.map(({id,label,building,floor,aliases})=>({id,label,building,floor:floor??1,aliases:aliases??[]}))}));
      if(pathname==='/favicon.ico') {res.writeHead(204);res.end();return;}
      const files:Record<string,[string,string]>={
        '/':['web/index.html','text/html; charset=utf-8'], '/evacuation':['web/evacuation.html','text/html; charset=utf-8'],
        '/evacuation/':['web/evacuation.html','text/html; charset=utf-8'], '/app.js':['web/app.js','text/javascript; charset=utf-8'],
        '/evacuation.js':['web/evacuation.js','text/javascript; charset=utf-8'], '/style.css':['web/style.css','text/css; charset=utf-8'],
        '/map.png':['src/map/gunn_site_map.png','image/png'], '/evacuation-map.png':['src/map/gunn_evacuation_map.png','image/png']};
      let file=files[pathname];
      if(/^\/output\/period_map(?:_[0-9a-f]{32})?\.png$/.test(pathname)) file=[resolve(outputDir,pathname.split('/').pop()!),'image/png'];
      if(!file) return send(res,404,JSON.stringify({error:'Not found'}));
      try {send(res,200,await readFile(resolve(ROOT,file[0])),file[1]);}
      catch(error) {if((error as NodeJS.ErrnoException).code==='ENOENT') send(res,404,JSON.stringify({error:'Not found'}));else throw error;}
    } catch(error) {
      if(!res.headersSent) send(res,error instanceof InputError?400:500,JSON.stringify({error:error instanceof InputError?error.message:'Unable to generate map'}));
      else res.end();
      if(!(error instanceof InputError)) console.error(error);
    }
  });
}
if(process.argv[1] && import.meta.url===pathToFileURL(resolve(process.argv[1])).href) {
  const portIndex=process.argv.indexOf('--port');
  const port=Number(portIndex>=0?process.argv[portIndex+1]:process.env.PORT??8000);
  if(!Number.isInteger(port)||port<0||port>65535) throw new Error('Invalid port');
  const host=process.env.HOST??'127.0.0.1';
  createApp().listen(port,host,()=>console.log(`Open http://${host}:${port}/`));
}
